pkg/manager/volumes/vol_compare_utils.go

```diff
@@ -0,0 +1,385 @@
+	}
+
+	ctx.pods = pods
+	ctx.sts = sts
+	ctx.shouldEvict = comp == v1alpha1.TiKVMemberType
+
+	return ctx, nil
+}
+
+func (u *volCompareUtils) getPodsOfComponent(tc *v1alpha1.TidbCluster, mt v1alpha1.MemberType) ([]*corev1.Pod, error) {
+	selector, err := u.sf.NewSelector(tc.GetInstanceName(), mt)
+	if err != nil {
+		return nil, err
+	}
+
+	ns := tc.GetNamespace()
+
+	pods, err := u.deps.PodLister.Pods(ns).List(selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list Pods: %w", err)
+	}
+
+	sort.Slice(pods, func(i, k int) bool {
+		a, b := pods[i].Name, pods[k].Name
+		if len(a) != len(b) {
+			return len(a) < len(b)
+		}
+		return a < b
+	})
+
+	return pods, nil
+}
+
+func (u *volCompareUtils) getStsOfComponent(cluster v1alpha1.Cluster, mt v1alpha1.MemberType) (*appsv1.StatefulSet, error) {
+	ns := cluster.GetNamespace()
+	stsName := controller.MemberName(cluster.GetName(), mt)
+
+	sts, err := u.deps.StatefulSetLister.StatefulSets(ns).Get(stsName)
+	if err != nil {
+		return nil, fmt.Errorf("get sts %s/%s failed: %w", ns, stsName, err)
+	}
+
+	return sts, nil
+}
+
+func (u *volCompareUtils) IsStatefulSetSynced(ctx *componentVolumeContext, sts *appsv1.StatefulSet, errOnNewVol bool) (bool, error) {
+	var newVolErr error = nil
+	for _, volTemplate := range sts.Spec.VolumeClaimTemplates {
+		volName := v1alpha1.StorageVolumeName(volTemplate.Name)
+		size := getStorageSize(volTemplate.Spec.Resources.Requests)
+		desired := getDesiredVolumeByName(ctx.desiredVolumes, volName)
+		if desired == nil {
+			// TODO: verify behaviour is correctly intended, this was previously being ignored. Now explicit error in
+			// pvc modifier..
+			errStr := fmt.Sprintf("volume %s in sts for cluster %s does not exist in desired volumes", volName, ctx.ComponentID())
+			if errOnNewVol {
+				newVolErr = errors.New(errStr)
+			} else {
+				klog.Warningf(errStr)
+			}
+			return false, newVolErr
+		}
+		if size.Cmp(desired.Size) != 0 {
+			return false, nil
+		}
+		scName := volTemplate.Spec.StorageClassName
+		if !isStorageClassMatched(desired.StorageClass, ignoreNil(scName)) {
+			return false, nil
+		}
+	}
+	// TODO: verify new behaviour is fine with pvc_modifier, previously number of volumes not checked.
+	if len(sts.Spec.VolumeClaimTemplates) != len(ctx.desiredVolumes) {
+		errStr := fmt.Sprintf("Number of volumes mismatch desired %d vs sts %d for cluster %s", len(ctx.desiredVolumes), len(sts.Spec.VolumeClaimTemplates), ctx.ComponentID())
+		if errOnNewVol {
+			newVolErr = errors.New(errStr)
+		} else {
+			klog.Warningf(errStr)
+		}
+		return false, newVolErr
+	}
+
+	return true, nil
+}
+
+func (u *volCompareUtils) IsPodSyncedForReplacement(ctx *componentVolumeContext, pod *corev1.Pod) (bool, error) {
+	// Does not check underlying StorageClass for modifications, only matches PVC specs for use by pvc replacer.
+	ns := pod.Namespace
+	podPvcCount := 0
+	for i := range pod.Spec.Volumes {
+		vol := &pod.Spec.Volumes[i]
+		pvc, err := u.getPVC(ns, vol)
```

Comment:
```suggestion
	for _, vol := range pod.Spec.Volumes {
		pvc, err := u.getPVC(ns, &vol)
```

Reply:
Applied.

Reply:
Actually had to undo this, I had to use i to avoid this linter error:

```
pkg/manager/volumes/vol_compare_utils.go:183:28: G601: Implicit memory aliasing in for loop. (gosec)
		pvc, err := u.getPVC(ns, &vol)
```

pointer to loop variable Issue might not apply in this case, but safer to avoid and appease the linter.